reporter.go

```diff
@@ -16,20 +16,20 @@ const (
 )
 
 type Reporter struct {
-	measurementSet *MeasurementSet
-	client         Client
-	prefix         string
+	measurementSet      *MeasurementSet
+	measurementsService *MeasurementsService
+	prefix              string
 
 	batchChan             chan *MeasurementsBatch
 	measurementSetReports chan *MeasurementSetReport
 
 	globalTags map[string]string
 }
```

Comment:
`outputMeasurementsIntervalSeconds` should probably be moved into the Reporter, where it could be used to set the flush interval in `flushReportsForever` and also could be passed into the MeasurementBatch as the `Period`.

Relatedly, I eventually changed the logic of `flushReportsForever` to this:

``` golang
func (r *Reporter) flushReportsForever() {
	for {
		// Sleep until the top of the next interval
		now := time.Now()
		nextInterval := now.Truncate(outputMeasurementsInterval).Add(outputMeasurementsInterval)
		time.Sleep(nextInterval.Sub(now))
		report := r.measurementSet.Reset()
		r.flushReport(report)
	}
}
```

This change aligns the measurement batch flushes to the start of each Period, rather than the current algorithm which randomizes the flush offset. In particular, if you have many measurements reporting independently at different offsets, the "last" value in metrics charts -- specifically noticeable in the Big Number UI widgets -- will generally reflect a random subset of the reporters and thus provide bogus/misleading data.

Reply:
@tillberg this is great feedback -- I really appreciate it!

I'll have to consider this in the context of the native batching constructs I'm building in as part of #4. Just FYI, my goal is to cover two main use cases w/ this client:

1. **instrumentation case**: as a developer instrumenting an application, I can set up a collection of Guages and Counters, update them in application code, and have the client automatically send results to AO (basically what `Reporter` helps support right now).

2. **entity direct creation case**: as a developer creating an integration (e.g. a Prometheus storage adapter or an OpenCensus exporter), I can arbitrarily create `Measurement` types, place them on a write-only channel, and then let the client handle batching/persistence on a configurable interval.

In both of these cases, we need batched `Measurement` persistence. If you look at some of the code that has been committed on #4, you'll see the approach I'm taking. I'm not 100% happy with it yet b/c it doesn't support everything you can do w/ the batching implementation you wrote in `Reporter`.  Generally though I'm hoping to push the batching concerns into one `BatchPersister` that can handle the batching needs of both of the above cases. That would mean excising functionality from `Reporter` while preserving the benefit of your work here vis-a-vis manipulating periods, aligning time intervals, etc.

Does all that make sense? Since I'm working on the batching code now, I thought it best to let you know the direction I'm currently thinking of taking.

Reply:
OK, yeah, that sounds great. It'll be nice to cover both those use cases well! In the latter case, it makes sense to push time/period concerns to the user.

Reply:
@tillberg `MultiReporter` has a similar randomized version of `flushReportsForever()`. Placing the interval on the `Reporter` means that it's not available to be calculated globally on that construct. Any objection to me just delegating the "forever" part to the individual `Reporter` types?

Reply:
@trevrosen ah, `MultiReporter` is new to me.

I don't think it will work to have each separate downstream `*Reporter` separately call `m.flushReport(m.measurementSet.Reset())` on the same MeasurementSet `m`, at least not with the way that we format reports by calling `Reset()` on the MeasurementSet: `Reset()` collects the full MeasurementSetReport by grabbing the values stored in each counter/gauge and resetting them to 0 in the same step and with only a single lock/unlock for each SynchronizedCounter/SynchronizedGauge.

Even if that method were changed to make it possible for each to independently flush, it might still be desirable (for users) to generate a single report and pass it to both of the downstream Reporters, as this will ensure that the metrics passed to each will be identical and not slightly misaligned from one another; that may or may not matter much depending on the use case, but I can see users sending data to two metrics pipelines and then wondering why the numbers are a little different in each.